src/resolver.rs

```diff
@@ -349,23 +368,57 @@ impl<'ctx> DependencyResolver<'ctx> {
         }
 
         // Mark all other versions of the dependency as invalid.
-        match src.state {
+        let new_ids = match src.state {
             State::Open => unreachable!(),
             State::Locked(_) => unreachable!(), // TODO: This needs to do something.
-            State::Constrained(ref mut ids) |
-            State::Picked(_, ref mut ids) => {
-                *ids = (*ids).intersection(&indices).map(|i| *i).collect();
-                if ids.is_empty() {
+            State::Constrained(ref ids) |
+            State::Picked(_, ref ids) => {
+                let is_ids = (*ids).intersection(&indices).map(|i| *i).collect::<HashSet<usize>>();
+                if is_ids.is_empty() {
                     let mut msg = format!("Requirement `{}` conflicts with other requirements on dependency `{}`.\n", con, name);
+                    let mut cons = Vec::new();
                     for &(pkg_name, ref con) in all_cons {
                         msg.push_str(&format!("\n- package `{}` requires `{}`", pkg_name, con));
+                        cons.push(con);
+                    }
+                    cons = cons.into_iter().unique().collect();
+                    println!("{}\n\nTo resolve this conflict manually, \
+                        select a revision for `{}` among:", msg, name);
+                    for (idx, e) in cons.iter().enumerate() {
+                        println!("{}) `{}`", idx, e);
+                    };
+                    print!("Enter a number or hit enter to abort: ");
+                    io::stdout().flush().unwrap();
+                    let mut buffer = String::new();
+                    io::stdin().read_line(&mut buffer).unwrap();
+                    let choice = buffer.trim().parse::<usize>().unwrap();
+                    let decision = cons[choice];
+                    match self.req_indices(name, decision, src) {
+                        Ok(o) => match o {
+                            Some(v) => Ok(v),
+                            None => Err(Error::new("something"))
+                        },
+                        Err(e) => Err(e)
                     }
-                    return Err(Error::new(msg));
+                } else {
+                    Ok(is_ids)
```

Comment:
Pretty elegant! I like the ability for the user to resolve conflicts. As you say in the PR, it would be great if we could wrap the line-reading in a loop that keeps asking the user for input as long as the input is invalid and the user hasn't hit enter.

Reply:
Done, see the new commits.